### PR TITLE
Revert "Revert "Merge pull request #50951 from ZhiguoZh/20230607-toyear-fix""

### PR DESCRIPTION
**Pull request by hanfei1991**
Try to use DayNum to avoid meeting DSL change boundary.
For example, America/Lima change DSL at 1994-01-01 00:00:00
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
convert condition like `toYear(x) = c` to `c1 <= x < c2`

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit ca6930eb110903709fc4c2e1cbec19a95e55ee18 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51795/ca6930eb110903709fc4c2e1cbec19a95e55ee18/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

**Comment by hanfei1991** [comment]
@ZhiguoZh how do you think of this change

**Comment by ZhiguoZh** [comment]
Hi @hanfei1991 , thanks so much for sharing your insights! And exactly, DST is the root cause and your solution successfully fixed the erroneous AST tests (as is shown in this [list](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gSU5URVJWQUwgMSBZRUFSCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyBMSUtFICdGJScKICAgIEFORCBjaGVja19zdGF0dXMgIT0gJ3N1Y2Nlc3MnCiAgICBBTkQgcG9zaXRpb24odGVzdF9uYW1lLCAnMDI3ODVfZGF0ZV9wcmVkaWNhdGVfb3B0aW1pemF0aW9uc19hc3RfcmV3cml0ZScpID4gMApPUkRFUiBCWSBjaGVja19zdGFydF90aW1l)) when the clickhouse-server is launched with DST-affected time zones, `America/Lima` and `America/Managua`.

My only concern is the return of a pair of `ExtendedDayNum` as the preimage would to some degree decrease the flexibility of the optimizer. For example, when we try to extend this optimization to `toStartOfFiveMinute`, whose preimage is a 5-minutes DateTime interval, the granularity of "day" should be insufficient in this case.

Considering that it's our purpose to generate strings of Date/DateTime which are not affected by the DST, how about explicitly specifying the DST-insensitive timezone, such as UTC, when acquiring the `DateLUTImpl` instance? The experiment shows that it also works for the `America/Lima` and `America/Managua` time zones.

Thanks again for figuring out the cause! It really helps me a lot!

File: src/Functions/DateTimeTransforms.h [hunk 1]
```diff
@@ -718,6 +723,28 @@ struct ToYearImpl
...
+    static inline constexpr bool hasPreimage() { return true; }
+
+    static inline RangeOrNull getPreimage(const IDataType & type, const Field & point)
+    {
+        if (point.getType() != Field::Types::UInt64) return std::nullopt;
+
+        auto year = point.get<UInt64>();
+        if (year < DATE_LUT_MIN_YEAR || year >= DATE_LUT_MAX_YEAR) return std::nullopt;
+
+        const DateLUTImpl & date_lut = DateLUT::instance();
```
Comment (ZhiguoZh) [inline src/Functions/DateTimeTransforms.h:735]: Replace with `const DateLUTImpl & date_lut = DateLUT::instance("UTC");`

File: src/Interpreters/OptimizeDateOrDateTimeConverterWithPreimageVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+ *  f(x) >  c -> x >= e
+ *  f(x) >= c -> x >= b
+ *  f(x) <  c -> x <  b
+ *  f(x) <= c -> x <  e
+ *
+ *  This function generates a new AST with the transformed relation.
+ */
+ASTPtr generateOptimizedDateFilterAST(const String & comparator, const NameAndTypePair & column, const std::pair<Field, Field>& range)
+{
+    const DateLUTImpl & date_lut = DateLUT::instance();
```
Comment (ZhiguoZh) [inline src/Interpreters/OptimizeDateOrDateTimeConverterWithPreimageVisitor.cpp:38]: The same here. `const DateLUTImpl & date_lut = DateLUT::instance("UTC");`

**Comment by hanfei1991** [comment]
yes, LUT utc is better ...

**Review by rschu1ze — APPROVED** [review]

**Pull request merged**